plaidml/bridge/openvino/ops/deformable_convolution.cpp

```diff
@@ -0,0 +1,284 @@
+  std::vector<int> OFF_broadcast, OFF_bcast_axes;
+  OFF_broadcast.push_back(N);
+  OFF_broadcast.push_back(DG);
+  OFF_broadcast.push_back(CI / DG);
+  OFF_broadcast.insert(OFF_broadcast.end(), NEW_DIM.begin(), NEW_DIM.end());
+  OFF_broadcast.push_back(rank - 2);
+  for (auto i = 0; i < rank + 2; ++i) {
+    OFF_bcast_axes.push_back(i);
+  }
+  offset = op::broadcast(offset, OFF_broadcast, OFF_bcast_axes);
+  // Set up the value of the reshape to multiply the channel dimension.
+  std::vector<int> OFF_reshape_channel;
+  OFF_reshape_channel.push_back(N);
+  OFF_reshape_channel.push_back(CI);
+  OFF_reshape_channel.insert(OFF_reshape_channel.end(), NEW_DIM.begin(), NEW_DIM.end());
+  OFF_reshape_channel.push_back(rank - 2);
+  offset = op::reshape(offset, make_tuple<int>(OFF_reshape_channel));  // the shape of offset is {N, CI, NEW_DIM,
+                                                                       // rank-2}
+  // Define the index of input.
+  std::vector<edsl::Tensor> index_vec(rank - 2);
+  for (auto i = 0; i < rank - 2; ++i) {
+    edsl::Tensor index_vec_0 =
+        edsl::index({edsl::TensorDim(OFF_shape[i + 2]), edsl::TensorDim(F_shape[i + 2])}, static_cast<size_t>(0));
+    edsl::Tensor index_vec_1 =
+        edsl::index({edsl::TensorDim(OFF_shape[i + 2]), edsl::TensorDim(F_shape[i + 2])}, static_cast<size_t>(1));
+    index_vec[i] = index_vec_0 * strides[i] + index_vec_1 * dilations[i] - pad_befores[i];
+    // Set up the value of reshape to broadcast index.
+    std::vector<int> index_reshape;
+    for (auto j = 0; j < i + 2; ++j) {
+      index_reshape.push_back(1);
+    }
+    index_reshape.push_back(NEW_DIM[i]);
+    for (auto j = 0; j < rank - 2 - i; ++j) {
+      index_reshape.push_back(1);
+    }
+    index_vec[i] = op::reshape(index_vec[i], make_tuple<int>(index_reshape));
+    // Set up the value of broadcast.
+    std::vector<int> result_shape;
+    std::vector<int> broadcast_axes;
+    result_shape.push_back(1);
+    result_shape.push_back(1);
+    broadcast_axes.push_back(0);
+    broadcast_axes.push_back(1);
+    for (auto j = 0; j < rank - 2; ++j) {
+      result_shape.push_back(NEW_DIM[j]);
+      broadcast_axes.push_back(j + 2);
+    }
+    result_shape.push_back(1);
+    broadcast_axes.push_back(rank);
+    index_vec[i] = op::broadcast(index_vec[i], result_shape, broadcast_axes);
+  }
+  edsl::Tensor index = op::concatenate(index_vec, -1);  // The shape of index is {1, 1, NEW_DIM, rank-2}
+  // Get deformabled index.
+  edsl::Tensor new_index = offset + index;
+  // Get deformable input tensor.
+  edsl::Tensor deform_input =
+      edsl::gather(I, new_index).axis(2 - rank).interpolationMode(edsl::InterpolationMode::LINEAR);
+  for (auto i = 1; i < rank - 2; ++i) {
+    deform_input =
+        edsl::gather(deform_input, new_index).axis(2 - rank + i).interpolationMode(edsl::InterpolationMode::LINEAR);
+  }
+  deform_input = extract_tensor(deform_input, rank);
+  /*
+  // Another method with gather_nd.
```

Comment:
You can delete this comment, I think gather ND is not going to support non-integer indices in the near future.

Reply:
ok, got it